bin/p2-label/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/flags"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/util"
+	"gopkg.in/alecthomas/kingpin.v2"
+	klabels "k8s.io/kubernetes/pkg/labels"
+)
+
+var (
+	cmdApply               = kingpin.Command(CmdApply, "Apply label changes to all objects matching a selector")
+	applyLabelType         = cmdApply.Flag("labelType", "The type of label to adjust. Sometimes called the \"label tree\"").Short('t').Required().String()
+	applySubjectSelector   = cmdApply.Flag("selector", "The selector on which to modify labels.").Short('s').Required().String()
+	applyAddititiveLabels  = cmdApply.Flag("add", "The label set to apply to the subject.").Short('a').String()
+	applyDestructiveLabels = cmdApply.Flag("delete", "The label set to remove to the subject. It is not an error to include extra labels here.").Short('d').String()
+
+	cmdShow       = kingpin.Command(CmdShow, "Show labels that apply to a particular entity (type, ID)")
+	showLabelType = cmdShow.Flag("labelType", "The type of label to adjust. Sometimes called the \"label tree\"").Short('t').Required().String()
+	showID        = cmdShow.Flag("id", "The ID of the entity to show labels for.").Short('i').Required().String()
+)
+
+const (
+	CmdApply = "apply"
+	CmdShow  = "show"
+)
+
+func main() {
+	cmd, opts := flags.ParseWithConsulOptions()
+	client := kp.NewConsulClient(opts)
+	applicator := labels.NewConsulApplicator(client, 3)
+
+	switch cmd {
+	case CmdShow:
+		labelType, err := labels.AsType(*showLabelType)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing label type. Check the commandline. \n%v\n", err)
+		}
+
+		labelsForEntity, err := applicator.GetLabels(labelType, *showID)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Got error while querying labels. %v\n", err)
+			return
+		}
+		fmt.Printf("The current labels for %s are: %s\n", *showID, labelsForEntity.Labels.String())
+		return
+	case CmdApply:
+		labelType, err := labels.AsType(*applyLabelType)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing label type. Check the commandline. \n%v\n", err)
+		}
+
+		subject, err := klabels.Parse(*applySubjectSelector)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error while parsing subject label. Check the syntax. \n%v\n", err)
+		}
+
+		additive, err := klabels.Parse(*applyAddititiveLabels)
```

Comment:
Kinpin has a better option than hacking around the klabel selector parser to get a label set. You can make the flag a `StringMap`, and it will automatically allow you to use the syntax `--add foo=bar --add biz=bam` and assign those to a map `map[string]string{"foo":"bar","biz":"bam"}`